src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.InFlightSolution.cs

```diff
@@ -0,0 +1,141 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    internal sealed partial class RemoteWorkspace
+    {
+        /// <summary>
+        /// Wrapper around asynchronously produced solution for a particular <see cref="SolutionChecksum"/>.  The
+        /// computation for producing the solution will be canceled when the number of in-flight operations using it
+        /// goes down to 0.
+        /// </summary>
+        public sealed class InFlightSolution
+        {
+            private readonly RemoteWorkspace _workspace;
+
+            public readonly Checksum SolutionChecksum;
+
+            private readonly CancellationTokenSource _cancellationTokenSource = new();
+
+            /// <summary>
+            /// Background work to just compute the disconnected solution associated with this <see cref="SolutionChecksum"/>
+            /// </summary>
+            private readonly Task<Solution> _disconnectedSolutionTask;
+
+            /// <summary>
+            /// Optional work to try to elevate the solution computed by <see cref="_disconnectedSolutionTask"/> to be
+            /// the primary solution of this <see cref="RemoteWorkspace"/>.  Must only be read/written while holding
+            /// <see cref="RemoteWorkspace._gate"/>.
+            /// </summary>
+            private Task<Solution>? _primaryBranchTask;
+
+            /// <summary>
+            /// Initially set to 1 to represent the operation that requested and is using this solution.  This also
+            /// allows us to use 0 to represent a point that this solution computation is canceled and can not be
+            /// used again.
+            /// </summary>
+            public int InFlightCount { get; private set; } = 1;
+
+            public InFlightSolution(
+                RemoteWorkspace workspace,
+                Checksum solutionChecksum,
+                Func<CancellationToken, Task<Solution>> computeDisconnectedSolutionAsync)
+            {
+                _workspace = workspace;
+                SolutionChecksum = solutionChecksum;
+
+                _disconnectedSolutionTask = computeDisconnectedSolutionAsync(_cancellationTokenSource.Token);
+            }
+
+            /// <summary>
+            /// Allow the RemoteWorkspace to try to elevate this solution to be the primary solution for itself.  This
+            /// commonly happens because when a change happens to the host, features may kick off immediately, creating
+            /// the disconnected solution, followed shortly afterwards by a request from the host to make that same
+            /// checksum be the primary solution of this workspace.
+            /// </summary>
+            /// <param name="updatePrimaryBranchAsync"></param>
+            public void TryKickOffPrimaryBranchWork_NoLock(Func<Solution, CancellationToken, Task<Solution>>? updatePrimaryBranchAsync)
```

Comment:
It seems like theres only ever one thing passed to this which is the function to update the remote workspace current solution.  Can that be inlined here so we could just pass a bool?

Reply:
ok.  i think i made this simpler :)